fixes #2130 - normalize_hostname appends domain to fqdn

frimik opened the pull request:
Ok, let me know if it makes sense - we assume otherwise that hostnames are RFC952 , right? Ie, dot (.) is only allowed as delimiter between domain components.

Commit description:

When the domain part of $fqdn doesn't match $domain, the normalize_hostname
function would append $domain onto the end of $fqdn, saving the host as
$fqdn.$domain (Example: hostname.domain1.com.domain2.com).

GregSutcliffe commented:
@frimik I cannot replicate the problem. I create a host called "foo.my.domain", and then changed it's domainname to "your.domain". so now it has a certname of "foo.my.domain" and a hostname of "foo.your.domain". When I get that host to report in to Foreman, Foreman just creates a new host for "foo.your.domain". I've tried this with managed and unmanaged Hosts.

Do you have a step-by-step reproduction?

frimik commented:
we discussed this on IRC. The issue is not new reported hosts. The issue is Editing an existing host which has a "mismatch" between fqdn and @host.domain. It will update the host to a bogus fqdn when you click submit.

ohadlevy commented:
@frimik sorry that it took me a while, can you add a couple of tests, so we can understand how this scenario happens?

thanks!

frimik commented:
I can try :)

frimik committed:
fixes #2130 - normalize_hostname appends domainname to fqdn

When the trailing part of $fqdn doesn't match $domain, the normalize_hostname
function would append $domain onto the end of $fqdn, saving the host as
$fqdn.$domain (Example: hostname.domain1.com.domain2.com).

This rename would happen if you edit a host already named 'host.domain1.com'
where a (hardcoded) certname says 'host.domain1.com' but the domain fact says
'domain2.com'.

If a hostname is supposed to follow RFC952 format, any . (dot) is only allowed
as a separator between domain parts and _not_ in the hostname part itself.
Thus if there is already a dot in the hostname we have to assume the host name
(fqdn) is already fully qualified.

domcleal commented:
[test]

rrtj3 commented:
I run into this same problem when I edit an existing host, setting a host-level parameter on the host. The hostname then appends the domain name to the host.

For example, 
- I have the certname hard-coded on the clients to: hostname.pvt.hawaii.edu.
- The hostname of the box is hostname.its.hawaii.edu
- I add a parameter to the host in foreman and save.
- It changes the name of the host to hostname.pvt.hawaii.edu.its.hawaii.edu

I tested the provided fix and it works well for me.

ohadlevy commented:
merged, @frimik apologies for taking so much time
